src/components/Rockets.js

```diff
@@ -0,0 +1,81 @@
+import React from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { bookRocket, cancelBooking } from '../redux/rockets/rocketsSlice';
+import './rockets.css';
+
+const Rockets = () => {
+  const dispatch = useDispatch();
+  const rockets = useSelector((state) => state.rockets.rockets);
+  const isLoading = useSelector((state) => state.rockets.isLoading);
+  const error = useSelector((state) => state.rockets.error);
+
+  const handleBookRocket = (rocketId) => {
+    dispatch(bookRocket(rocketId));
+  };
+
+  const handleCancelBooking = (rocketId) => {
+    dispatch(cancelBooking(rocketId));
+  };
+
+  if (isLoading) {
+    return (
+      <>
+        <p>loading...</p>
+        <div className="Rockets-preloader" />
+      </>
+    );
+  }
+
+  if (error) {
+    return (
+      <>
+        <p>unable to load the Rockets...</p>
+        <p>
+          Error:
+          {error}
+        </p>
+      </>
+    );
+  }
+
+  return (
+    <div>
+      <h2 className="h2-rockets">Rockets</h2>
+      <div title="rockets-component" className="Rockets">
+        {rockets.map((rocket) => {
+          const isBooked = rocket.reserved;
+          const handleClick = isBooked ? handleCancelBooking : handleBookRocket;
+          const badgeDisplayed = isBooked
+            ? 'Rocket-display-badge'
+            : 'Rocket-no-display-badge';
+
+          return (
+            <div key={rocket.id} className="Rockets-container">
+              <img
+                src={rocket.flickr_images[0]}
+                alt={rocket.name}
+                className="Rocket-img"
+              />
+              <div className="Rockets-content">
+                <h2>{rocket.name}</h2>
+                <p>
+                  <span className={badgeDisplayed}>Booked</span>
+                  {rocket.description}
+                </p>
+                <button
+                  type="submit"
+                  className={isBooked ? 'cancel-button' : 'book-button'}
+                  onClick={() => handleClick(rocket.id)}
+                >
+                  {isBooked ? 'Cancel Reservation' : 'Reserve Rocket'}
+                </button>
```

Comment:
- [ Required ] I think it would be better to use conditional components rendering. It's considered good practice not to handle app logic of any kind in a React component, a good React component should only be used for display purposes. I would like to suggest that you refactor this code to use conditional components as shown below.

```JavaScript
   return (
    <div>
      <h2 className="h2-rockets">Rockets</h2>
      <div title="rockets-component" className="Rockets">
        {rockets.map((rocket) => (
          <div key={rocket.id} className="Rockets-container">
            <img
              src={rocket.flickr_images[0]}
              alt={rocket.name}
              className="Rocket-img"
            />
            <div className="Rockets-content">
              <h2>{rocket.name}</h2>
              <p>
                {(rocket.reserved
                && <span className="Rocket-display-badge">Booked</span>
                )}
                {rocket.description}
              </p>
              {(rocket.reserved

              && (
              <button
                type="submit"
                className="cancel-button"
                onClick={() => handleCancelBooking(rocket.id)}
              >
                Cancel Reservation
              </button>
              ))}

              {(!rocket.reserved
              && (
              <button
                type="submit"
                className="book-button"
                onClick={() => handleBookRocket(rocket.id)}
              >
                Reserve Rocket
              </button>
              ))}
            </div>
          </div>
        ))}
      </div>
    </div>
  );
```

## Project requirements
![Screenshot from 2023-04-14 23-37-38](https://user-images.githubusercontent.com/16392046/232150341-3c73e49b-d0aa-4b0a-b1cc-9d64386dd1bf.png)

Reply:
Thanks for your keen observation, the requested change is implemented